### ISSUE TITLE
Bug: [@typescript-eslint/no-unnecessary-condition] false positive with optional chaining (TS 5.3.2)

### ISSUE THREAD

**Issue by vlinder** [typescript-eslint/typescript-eslint#7977]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.ts&code=FDCmA8AcHsCcBcAEBjaA7Azk%2BouILyIAUAhgPwBciA3ogEZV3TQA2oJaiAvgJSPNsOBAHw1giFOjzICicgDo6BfIQBmJFhlCIyiAIyIqAJgDc4yZiQATWQqUq1GrTv2HEp87FDwArrE4yolZmXEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gdQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
export const test = (a?: { b: boolean }): boolean => {
  const c = a?.b === false ? 1 : 2;
  const d = a?.b === false ? 1 : 2;
  return c > d;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

There should be no linting error for the expression of variable `d`.

### Actual Result

It reports no-unnecessary-condition.

### Additional Info

This became an issue after we updated to typescript 5.3.2 from 5.2.2.

**Comment by bradzacher** [comment]
The rule is powered by types.
In this case TS has narrowed the type of `a` to not be undefined.
[playground](https://www.typescriptlang.org/play?#code/FDCmA8AcHsCcBcAEBjaA7Azk+ouILyIAUAhgPwBciA3ogEZV3TQA2oJaiAvgJSPNsOBAHw1giFOjzICicgDo6BfIQBmJFhlCIyiAIyIqAJgDc4yZiSrmshXTMSA9I4muAemXOpLiACa2yRWU1DS0dfUNEU3NzWFB4AFdYThlRXzMuYCA)

So the rule reports because according to the types it is not `undefined`.

**Comment by vlinder** [comment]
Thanks, I see that they fixed this in 5.4 nightly..
